### DISCUSSION TITLE
Does Gulp v4 eliminate need for pump or pipeline?

### DISCUSSION THREAD

**Discussion by rdhallman** [gulpjs/gulp#2530] [Help] [answered]
Several articles on Gulp best practices advocate for using `pump()` instead of `.pipe()`.  For example:

https://github.com/gulpjs/gulp/tree/master/docs/why-use-pump
https://django.cowhite.com/blog/using-gulp-to-pipeline-front-end-builds/

However, articles like this are from the Gulp 3 days.  Is such advice still relevant given the "ground-up rewrite" of Gulp 4?

And if `pump()` is still good practice with Gulp 4, should `pump()` now be replaced with `pipeline()`:

https://stackoverflow.com/questions/58875655/whats-the-difference-between-pipe-and-pipeline-on-streams/60459320#60459320

**Comment by yocontra** [comment, marked as answer]
@rdhallman `pipeline` can be used like so:

```js
import { pipeline } from 'stream'

export const someTask = (done) => {
  pipeline(a, b, c, d, done)
}
```

If you use it a lot you could also `promisify` it via the node util so you could just return it and save some syntax.

pipeline vs. pump I think is just personal preference. The way `pipeline` works was based off of `pump` / `pumpify` (mafintosh projects) so the behavior should be nearly identical - I do know from personal experience that `pipeline` has some edge cases but I think they are resolved as of node 15.

**Reply by phated** [reply]
One slight update about `pump` vs `pipeline` - while `pipeline` can be returned from a task, the `pump` module must use the `done` callback. This is because the `pipeline` method in node will propagate an error to the stream, but `pump` won't.
